src/system_database.ts

```diff
@@ -113,17 +114,27 @@ export class PostgresSystemDatabase implements SystemDatabase {
   async flushWorkflowStatusBuffer() {
     const localBuffer = new Map(this.workflowStatusBuffer);
     this.workflowStatusBuffer.clear();
-    const client: PoolClient = await this.pool.connect();
-    await client.query("BEGIN");
-    for (const [workflowUUID, output] of localBuffer) {
-      await client.query(
-        `INSERT INTO workflow_status (workflow_uuid, status, output) VALUES($1, $2, $3) ON CONFLICT (workflow_uuid)
-        DO UPDATE SET status=EXCLUDED.status, output=EXCLUDED.output;`,
-        [workflowUUID, StatusString.SUCCESS, JSON.stringify(output)]
-      );
+    try {
+      const client: PoolClient = await this.pool.connect();
+      await client.query("BEGIN");
+      for (const [workflowUUID, output] of localBuffer) {
+        await client.query(
+          `INSERT INTO workflow_status (workflow_uuid, status, output) VALUES($1, $2, $3) ON CONFLICT (workflow_uuid)
+          DO UPDATE SET status=EXCLUDED.status, output=EXCLUDED.output;`,
+          [workflowUUID, StatusString.SUCCESS, JSON.stringify(output)]
+        );
+      }
+      await client.query("COMMIT");
+      client.release();
+    } catch (error) {
+      this.logger.error("Error flushing workflow buffer", error)
+      // If there is a failure in flushing the buffer, return items to the global buffer for retrying later.
+      for (const [workflowUUID, output] of localBuffer) {
+        if (!this.workflowStatusBuffer.has(workflowUUID)) {
+          this.workflowStatusBuffer.set(workflowUUID, output)
+        }
+      }
```

Comment:
Maybe useful: we could timestamp buffer entries so we can check how long they've been there. (Maybe there is an easier/ more elegant way to observe this.)